proposals/pending-state.md

```diff
@@ -0,0 +1,105 @@
+
+It's often desirable to communicate the state of async work up the component tree to parent components so that they can display user affordances indicating whether the UI or content is pending some async operation.
+
+Frameworks have the ability to invent custom APIs and patterns to handle this kind of cross-component communication. To enable the same with web components in a way that's interoperable between components from diffrent sources, and possibly implemented with different libraries, we need a protocol that components can implement without reliance on a common implementation.
+
+# Goals
+
+* Allow components to communicate that they have pending asynchonous work to ancestors in the DOM tree.
+* Allow components to know if there is pending asynchronous work in descendants in the DOM tree.
+* Allow components to intercept, modify, and block pending state notifications.
+* Give guidance on the type of asynchronous work that should be notified.
+* Allow for some labelling or differentiation between types of asynchronous work.
+
+# Details
+
+## Asynchronous States
+
+There are four main states that an asynchronous operation can be in:
+
+* Not-started
+* Started
+* Sucessfully completed
+* Failed
+
+## The `pending-state` event
+
+Components with pending work indicate so by firing a composed, bubbling, `pending-state` Event, with a `promise` property:
+
+TypeScript interface:
+```ts
+interface PendingStateEvent extends Event {
+  promise: Promise<void>;
+}
+```
+
+Example:
+
+```js
+class PendingStateEvent extends Event {
+  constructor(promise) {
+    super('pending-state', {bubbles: true, composed: true});
+    this.promise = promise;
+  }
+}
+
+// Inside a component definition:
+class DoWorkElement extends HTMLElement {
+  async doWork() { /* ... */ }
+
+  startWork() {
+    this.dispatchEvent(new PendingStateEvent(this.doWork()));
+  }
+}
+
+// Inside a container component:
+class IndicateWorkElement extends HTMLElement {
+  constructor() {
+    super();
+    this.addEventListener('pending-state', async (e) => {
+      this.showSpinner();
+      await e.promise;
+      this.hideSpinner();
+    });
+  }
+}
+```
+
+The Promise must be resolved when the work is complete and rejected if the work fails. The value the Promise resolves to is unspecified.
+
+Using an event with a Promise allows us to represent three of the four asynchronous states:
+
+* Not-started: not represented
+* Started: `pending-state` event fired
+* Sucessfully completed: Promise resolved
+* Failed: Promise rejected
+
+# Open Questions
+
+## Should the event carry a Promise or should there be two events?
+
+Animations have multiple events: `animationstart` and `animationend`. Promises make it very easy to correlate the start of an operation with the end of that operation.
+
+## Event Name
+
+Is `pending-state` good? Should it be `pending-work`?
```

Comment:
`pending-work` sounds better to me 🤗

Reply:
+1 pending-work -- "state" has connotations of redux and similar systems.

Reply:
Kabob casing this may align with custom element naming and its goal not to step on native things, but what if `pending` was used to further generalize this events relationship to `work`, `loading`, `rendering`, `progress`, etc?

Reply:
`pending-task` is another idea.

I'm a little way of such a short name like `pending`, just for increased collision chances.

Reply:
I renamed to `pending-task` for now. I think it's slightly better than `pending-work` and it felt right to replace instance of both "state" and "work" in the docs with "task".